[SelectionDAG] Handle more opcodes in canCreateUndefOrPoison

bjope opened the pull request:
[SelectionDAG] Handle more opcodes in canCreateUndefOrPoison

Handle SELECT_CC similarly as SETCC.

Handle these operations that only propagate poison/undef based on the
input operands:
  SADDSAT, UADDSAT, SSUBSAT, USUBSAT, MULHU, MULHS,
  SMIN, SMAX, UMIN, UMAX

These operations may create poison based on shift amount and exact
flag being violated:
  SRL, SRA

One goal here is to allow pushing freeze through these operations
when allowed, as well as letting analyses such as
isGuaranteedNotToBeUndefOrPoison to not break on such operations.

Since some problems have been observed with pushing freeze through
SRA/SRL we block that explicitly in DAGCombiner::visitFreeze now.
That way we can still model SRA/SRL properly in
SelectionDAG::canCreateUndefOrPoison, e.g. when used by
isGuaranteedNotToBeUndefOrPoison, even if we do not want to push
freeze through those instructions.

bjope committed:
[SelectionDAG] Handle more opcodes in canCreateUndefOrPoison

Handle SELECT_CC similarly as SETCC.

Handle these operations that only propagate poison/undef based on the
input operands:
  SADDSAT, UADDSAT, SSUBSAT, USUBSAT, MULHU, MULHS,
  SMIN, SMAX, UMIN, UMAX

These operations may create poison based on shift amount and exact
flag being violated:
  SRL, SRA

One goal here is to allow pushing freeze through these operations
when allowed, as well as letting analyses such as
isGuaranteedNotToBeUndefOrPoison to not break on such operations.

Since some problems have been observed with pushing freeze through
SRA/SRL we block that explicitly in DAGCombiner::visitFreeze now.
That way we can still model SRA/SRL properly in
SelectionDAG::canCreateUndefOrPoison, e.g. when used by
isGuaranteedNotToBeUndefOrPoison, even if we do not want to push
freeze through those instructions.

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -15383,6 +15383,12 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
 
+  // There is a reverse transform in visitEXTRACT_VECTOR_ELT, so we need to
+  // avoid inifite looping by not pulling freeze through EXTRACT_VECTOR_ELT
+  // here.
```
I would really like to remove that fold .....................

nikic commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -15383,6 +15383,12 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
     return N0;
 
+  // There is a reverse transform in visitEXTRACT_VECTOR_ELT, so we need to
+  // avoid inifite looping by not pulling freeze through EXTRACT_VECTOR_ELT
```
```suggestion
  // avoid infinite looping by not pushing freeze through EXTRACT_VECTOR_ELT
```

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5073,6 +5073,16 @@ bool SelectionDAG::canCreateUndefOrPoison(SDValue Op, const APInt &DemandedElts,
+  case ISD::SMIN:
+  case ISD::SMAX:
+  case ISD::UMIN:
+  case ISD::UMAX:
```
How many of these do we have test coverage for? One of the main reasons I stopped adding opcodes was because it was getting trickier to add tests.

nikic replied:
TBH I don't think there is a lot of value in explicitly testing all the opcodes, as long as they don't have any special logic.

bjope replied:
I haven't really investigated coverage for any of these. I detected that the analysis stopped on these downstream in a few cases, and thought that they were "safe to add". But maybe we want to have some sort of test coverage (although I'm not sure what kind of test that is preferred).
OTOH, I doubt that we have test coverage for all operations that currently just return true by default either ;-)

nikic commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5091,8 +5101,18 @@ bool SelectionDAG::canCreateUndefOrPoison(SDValue Op, const APInt &DemandedElts,
+    if (Op.getOperand(0).getValueType().isInteger())
+      return false;
+
+    // FIXME: Handle FP compares.
```
Can this be combined with the following SETCC case, which has FP handling?

bjope replied:
Hmm, yes, it is probably just as simple as that (if just making sure we get the cond code  from correct operand). I will updated the patch.

RKSimon commented on llvm/test/CodeGen/X86/freeze-binary.ll:
```diff
@@ -581,19 +581,41 @@ define i32 @freeze_lshr_exact(i32 %a0) nounwind {
+  ret i32 %z
+}
+
+define i32 @freeze_lshr_exact_extra_use(i32 %a0, ptr %escape) nounwind {
```
pre-commit this test so we see the diff

topperc commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5091,8 +5101,18 @@ bool SelectionDAG::canCreateUndefOrPoison(SDValue Op, const APInt &DemandedElts,
   case ISD::BITCAST:
   case ISD::BUILD_VECTOR:
   case ISD::BUILD_PAIR:
+  case ISD::SELECT:
```
Do we need to worry about Fast math flags on SELECT?

bjope replied:
Thanks! After several years I still did not know that `select` could have flags.
I really think that we shuold document such things also for ISD opcodes (at least in ISDOpcodes.h).

I'll try to add an assert here to make sure that none of the instructions in the group have flags. And then I'll move the handling of SELECT further down in this switch, to make sure that we consider the flags in some way.

RKSimon commented on llvm/test/CodeGen/RISCV/rv64xtheadbb.ll:
```diff
@@ -260,15 +260,17 @@ define i32 @ctlz_lshr_i32(i32 signext %a) {
 ; RV64I-NEXT:    or a0, a0, a1
-; RV64I-NEXT:    srliw a1, a0, 16
+; RV64I-NEXT:    slli a1, a0, 33
+; RV64I-NEXT:    srli a1, a1, 49
```
Same problem as when we tried to deal with SRA/SRL back on https://reviews.llvm.org/D136529#4120959

bjope commented:
I've rebased this patch (since for example the orignally planned changes releated to EXTRACT_VECTOR_ELT has landed via other patches).
- The new version is including some special limitations for SRA/SRL in DAGCombiner::visitFreeze, due to the problems seen in earlier attempts to handle those operations. The test cases I had planned for SRA/SRL was however pre-committed already in commit 4d0f79e346ceb0ddb25a9.
- I've dropped everything related to SELECT for now (but SELECT_CC is still handled).

nikic reviewed (approved):
LGTM

DavidSpickett commented:
As this updated some tests, I've had to revert it while reverting https://github.com/llvm/llvm-project/pull/85932. I've put the details on that PR.

If this is actually unrelated to that feel free to reland, I just wasn't going to to be able to resolve the test file conflict myself so I went with the safest option.

bjope commented:
> As this updated some tests, I've had to revert it while reverting #85932. I've put the details on that PR.
> 
> If this is actually unrelated to that feel free to reland, I just wasn't going to to be able to resolve the test file conflict myself so I went with the safest option.

Re-applied in https://github.com/llvm/llvm-project/commit/55c6bda01ef5a166a69b43956775272d9d67bda5
